src/main/java/io/cdap/delta/bigquery/BigQueryUtils.java

```diff
@@ -0,0 +1,130 @@
+import com.google.cloud.bigquery.QueryJobConfiguration;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.TableResult;
+import io.cdap.delta.api.DeltaTargetContext;
+import io.cdap.delta.api.SourceTable;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+
+public final class BigQueryUtils {
+  private static final int MAX_LENGTH = 1024;
+  // according to big query dataset and table naming convention, valid name should only contain letters (upper or
+  // lower case), numbers, and underscores
+  private static final String VALID_NAME_REGEX = "[\\w]+";
+  private static final String INVALID_NAME_REGEX = "[^\\w]+";
+
+  private BigQueryUtils() {
+  }
+
+  static long getMaximumExistingSequenceNumber(DeltaTargetContext context, String project, @Nullable String datasetName,
+                                               BigQuery bigQuery, EncryptionConfiguration encryptionConfiguration) {
+    StringBuilder builder = new StringBuilder();
+    builder.append("SELECT MAX(max_sequence_num) FROM (");
+    Set<SourceTable> allTables = context.getAllTables();
+    List<String> maxSequenceNumQueryPerTable = new ArrayList<>();
+    for (SourceTable table : allTables) {
+      TableId tableId = TableId.of(project,
+                                   datasetName != null ? normalize(datasetName) : normalize(table.getDatabase()),
+                                   normalize(table.getTable()));
+      if (bigQuery.getTable(tableId) != null) {
+        maxSequenceNumQueryPerTable.add(String.format("SELECT MAX(_sequence_num) as max_sequence_num FROM %s.%s.%s",
+                                                      tableId.getProject(), tableId.getDataset(), tableId.getTable()));
+      }
+    }
+
+    builder.append(String.join(" UNION ", maxSequenceNumQueryPerTable));
+    builder.append(");");
+
+    long maxSequenceNumber;
+    try {
+      maxSequenceNumber = maxSequenceNumQueryPerTable.size() == 0 ? 0 : executeMaxQuery(bigQuery, builder.toString(),
+                                                                                        encryptionConfiguration);
+    } catch (InterruptedException e) {
+      throw new RuntimeException("Failed to compute the maximum sequence number among all existing tables.", e);
+    }
+    return maxSequenceNumber;
+  }
+
+  static long getMaximumSequenceNumberForTable(BigQuery bigQuery, TableId tableId,
+                                               EncryptionConfiguration encryptionConfig) throws Exception {
+    if (bigQuery.getTable(tableId) == null) {
+      return 0L;
+    }
+
+    String query = String.format("SELECT MAX(_sequence_num) FROM %s.%s", tableId.getDataset(), tableId.getTable());
+    return executeMaxQuery(bigQuery, query, encryptionConfig);
+  }
+
+  private static long executeMaxQuery(BigQuery bigQuery, String query, EncryptionConfiguration encryptionConfig)
```

Comment:
nit: executeAggregateQuery

Here the function itself is not executing a MaxQuery, instead it's executing a query passed in which is expected to only generate an aggregated result (one row and one column)

Reply:
+1 fixed.